### PR TITLE
Deprecate --certificate-email flag. Make --certificate-identity and -…

### LINKED ISSUE #2056: Discourage (disallow?) naked `cosign verify` invocations
Inspired by https://github.com/sigstore/sigstore-python/issues/155

`cosign verify` (in keyless mode) without any `--cert-email` provided will succeed when literally anybody has signed an image/blob. This is almost never the right behavior: evil@hacker.com can sign any blob you can.

Instead, we should require *some* verification policy:

- `--cert-email` (*and* `--cert-oidc-provider`; see #1947)
- a CUE policy
- `--insecure`, which will do the current behavior today (possibly a hidden flag)

(Keyfull mode is okay, because you're passing the key material or CA in explicitly)

### PR DESCRIPTION
**Pull request by kpk47**
…-certificate-oidc-issuer required for verification.

--certificate-email is now an alias for --certificate-identity

Signed-off-by: kpk47 <kkris@google.com>

#### Summary

Closes #2056

### CODE REVIEW

File: cmd/cosign/cli/options/certificate.go
Comment (woodruffw) [inline cmd/cosign/cli/options/certificate.go:47]: This is a tiny thing, but is it possible to make these options mutually exclusive? Passing both of them at once implies user error/confusion, since only one takes effect at a time.

File: cmd/cosign/cli/options/certificate.go
Comment (haydentherapper) [inline cmd/cosign/cli/options/certificate.go:44]: I'd just remove this flag actually. This is going to be going into 2.0, so we can break things.

Comment (woodruffw) [reply cmd/cosign/cli/options/certificate.go:44]: +1 -- I was thinking in the context of sigstore-python where we'll remove a whole bunch of things right before 1.0, but removal right now makes much more sense in the context of cosign 2.0.

Comment (kpk47) [reply cmd/cosign/cli/options/certificate.go]: Done.

File: cmd/cosign/cli/options/certificate.go
Comment (haydentherapper) [inline cmd/cosign/cli/options/certificate.go:44]: Is there enforcement that both of these flags are provided?

File: cmd/cosign/cli/verify.go
Comment (haydentherapper) [inline cmd/cosign/cli/verify.go:96]: I think we instead want this in the Exec functions, like https://github.com/sigstore/cosign/blob/1c15b03a6e9e47f5c85a8fd7ff98bf75e910df6a/cmd/cosign/cli/verify/verify.go#L78

You also need this in the other verify commands, for manifest, dockerfile, and verify-blob

File: cmd/cosign/cli/options/certificate.go
Comment (znewman01) [reply cmd/cosign/cli/options/certificate.go]: Outdated comment, we're just axeing `--certificate-email` altogether

**Review by znewman01 — COMMENTED** [review]
Overall LGTM. Needs a rebase before we can merge.

I'd be interested in putting this validation all in the same place. Maybe factor out some of these arguments into a new `options.VerifyOpts` and have a function to parse them into `cosign/verify.CheckOpts`. But I'm okay doing that later.

**Review by znewman01 — DISMISSED** [review]
adding approving review so tests run

**Review by znewman01 — DISMISSED** [review]
approving for tests

**Comment by znewman01** [comment]
test failure looks related!

**Review by znewman01 — DISMISSED** [review]
approve for tests, ping me once you're ready for full review

**Comment by kpk47** [comment]
@znewman01 This is ready for review now. The biggest changes are in test/sign_blob_test.sh, where I had to switch the keyless entry to using a local key and drop the verification step for the self-signed certificate. I believe the test logic is still sound (sign and verify with cosign, upload signature with self-signed cert, check that the first entry is still verifiable).

File: cmd/cosign/cli/verify/verify_blob.go
Comment (haydentherapper) [inline cmd/cosign/cli/verify/verify_blob.go:106]: Can you add a test for failure if these aren't set?

File: pkg/cosign/verify.go
Comment (haydentherapper) [inline pkg/cosign/verify.go:251]: remove?

File: test/sign_blob_test.sh
Comment (haydentherapper) [inline test/sign_blob_test.sh:33]: What is the reason for needing to change this? Do we lose coverage of the keyless flow with this change? I think both are useful as e2e tests.

File: cmd/cosign/cli/verify/verify_blob.go
Comment (znewman01) [inline cmd/cosign/cli/verify/verify_blob.go:96]: I don't *love* the duplication here. Could we pull this out into a common "VerificationIdentity" struct that goes into the `VerifyBlobCmd` (and other `*BlobCmd` structs) and stick the check there?

File: cmd/cosign/cli/verify/verify.go
Comment (kpk47) [reply cmd/cosign/cli/verify/verify.go:96]: That makes sense to me, but how should it behave with a local bundle? IIUC, the bundle wraps the cert and additional metadata. Do we also need `--cert-identity` and `--cert-oidc-issuer` in that case?

Comment (znewman01) [reply cmd/cosign/cli/verify/verify.go:96]: >  Do we also need --cert-identity and --cert-oidc-issuer in that case?

IMO: Yes, we do.

We don't want to trust the identity from the bundle; we want to cross-reference it with the `--cert-*` flags.

File: cmd/cosign/cli/verify/verify_attestation.go
Comment (kpk47) [reply cmd/cosign/cli/verify/verify_attestation.go]: Yes, this was a typo.

File: cmd/cosign/cli/verify/verify_blob.go
Comment (kpk47) [reply cmd/cosign/cli/verify/verify_blob.go]: I moved the check into CheckVerifyOptions. WDYT? I'm not convinced it's the best solution, but it _is_ already referenced by all the Check*Cmd structs.

File: cmd/cosign/cli/verify/verify_blob.go
Comment (kpk47) [reply cmd/cosign/cli/verify/verify_blob.go]: Done.

File: pkg/cosign/verify.go
Comment (kpk47) [reply pkg/cosign/verify.go]: Done

File: test/sign_blob_test.sh
Comment (kpk47) [reply test/sign_blob_test.sh:33]: I may be able to keep this keyless depending on how GitHub OIDC mints tokens, though we'd need some extra work to make it work when run locally.

What's the intended usage for this test? I don't see it referenced in any workflows, so I added it to tests.yaml.

File: cmd/cosign/cli/verify/verify_blob.go
Comment (znewman01) [reply cmd/cosign/cli/verify/verify_blob.go]: I like this overall, I think. I'm finding it more readable.

I was imagining sticking the `KeyRef` inside `CertVerifyOptions` as well so we wouldn't have to duplicate the `if KeyRef != ""` check. Then, the `Identities()` would return basically a sum-type of `cosign.Identity | PublicKey` you could `switch` over that. WDYT?

Part of what you're running up against is that this whole thing needs a big refactor (not your responsibility), gross. Thanks for your patience :)

File: cmd/cosign/cli/verify/verify.go
Comment (znewman01) [reply cmd/cosign/cli/verify/verify.go]: I think this has made some e2e tests start failing.

File: cmd/cosign/cli/verify/verify_blob.go
Comment (kpk47) [reply cmd/cosign/cli/verify/verify_blob.go]: I'm not sure I like that solution since it's pulling (a tiny) part of key parsing into a struct that's about certificate verification.

I filed https://github.com/sigstore/cosign/issues/2462 with my suggestions for how to go about the larger refactor. How about we deal with the duplication in this PR and then I address the larger refactor in a follow-on change?

File: cmd/cosign/cli/verify/verify.go
Comment (znewman01) [reply cmd/cosign/cli/verify/verify.go]: Pinging this—I don't want the `--bundle` flag to turn off the requirement for a key or identity.

**Review by znewman01 — DISMISSED** [review]
approving for tests only

**Comment by haydentherapper** [comment]
Bumping PR, @kpk47, what is left? You'll need to rebase since we had to change import paths to include `v2`

**Comment by kpk47** [comment]
Sorry, this fell off my plate. There isn't any major work left here, just fixing up a broken integration test. I've got time blocked out to finish this in the next few days.

File: test/sign_blob_test.sh
Comment (kpk47) [reply test/sign_blob_test.sh:33]: I got this to work by treating the values of certificate-identity and certificate-oidc-issuer as regexps and then having the test pass in a match-all pattern.

File: cmd/cosign/cli/options/certificate.go
Comment (haydentherapper) [inline cmd/cosign/cli/options/certificate.go:92]: We don't need to set `IssuerRegExp` and `SubjectRegExp` here, since the case statement just ignores them if they're not set. 

I think we'd need another set of flags to support verifying by regex (for example, an email could contain a period), but I think we could add this later on. Given this is a CLI interface, I don't expect needing to verify a regex.

EDIT: I see you are leveraging these flags as a regex later on. Do we want to just go ahead and add flags for certificate-identity-regex and certificate-oidc-issuer-regex now?

File: pkg/cosign/verify.go
Comment (haydentherapper) [inline pkg/cosign/verify.go:307]: Same here, a certificate subject is not the same as the subject alternative name, and we don't expect fulcio certificates to ever set the certificate subject.

File: test/sign_blob_test.sh
Comment (haydentherapper) [inline test/sign_blob_test.sh:36]: As I noted above, I think we shouldn't overload this flag to support regex, and should add a separate flag. Otherwise, you run into potential risks with a character representing both regex and a certain character (like a period, valid for emails)

File: pkg/cosign/verify.go
Comment (kpk47) [reply pkg/cosign/verify.go:307]: Do we care about non-Fulcio certificates that don't include SANs? We're using such a certificate in test/sign_blob_test.sh, so I figured it must be a supported use case.

File: pkg/cosign/verify.go
Comment (kpk47) [reply pkg/cosign/verify.go:281]: Done

File: cmd/cosign/cli/options/certificate.go
Comment (kpk47) [reply cmd/cosign/cli/options/certificate.go:92]: Done. Added the new flags.

File: test/sign_blob_test.sh
Comment (kpk47) [reply test/sign_blob_test.sh:36]: Done

File: pkg/cosign/verify.go
Comment (haydentherapper) [reply pkg/cosign/verify.go]: I don't think we should. The bring-your-own-pki story needs to be fleshed out and we'd need to think through how to represent a subject (do we do something like `CN=foo;O=bar`, or do we support flags for each subject component?)

Do you want me to help regenerate a test certificate?

**Review by Hayden-IO — COMMENTED** [review]
Just need to rerun `make docgen`, this looks great!

File: pkg/cosign/verify.go
Comment (kpk47) [reply pkg/cosign/verify.go]: Yes, please.

**Comment by haydentherapper** [comment]
@kpk47 Can you update `test/testdata/test_blob_cert.pem` with the following:

```
-----BEGIN CERTIFICATE-----
MIIBdDCCARqgAwIBAgIUZw7gQ6T/IgmiMD1AWB2OTIIVH1owCgYIKoZIzj0EAwIw
ADAeFw0yMjEyMjEwMDIwNThaFw0zMjEyMTgwMDIwNThaMAAwWTATBgcqhkjOPQIB
BggqhkjOPQMBBwNCAAR1Q4hB1jtagrdsVxygtDa/rli00U7n/1I/NSw8yoMRQ+MO
AjRhg3gtcV0tha34L6150qJirQHbfocsao8X6wFmo3IwcDAdBgNVHQ4EFgQUx3Wb
0LwCWoGsl0FUpeQb3M4MukkwHwYDVR0jBBgwFoAUx3Wb0LwCWoGsl0FUpeQb3M4M
ukkwEgYDVR0TAQH/BAgwBgEB/wIBATAaBgNVHREEEzARgQ9mb29AZXhhbXBsZS5j
b20wCgYIKoZIzj0EAwIDSAAwRQIhALXG7XS5TIFLp+jLSxjuRk1Tj5MfE+y9x92Z
YPMbi9GZAiAmfEe0+q5l3PnI6zliOG5kG6EcS80QQgQmPcFvRZWOvw==
-----END CERTIFICATE-----

```

And update https://github.com/sigstore/cosign/blob/main/test/testdata/README.md, the openssl command, to be :

```
openssl req -key test/testdata/test_blob_private_key -x509 -days 3650 -out cert.pem -new -nodes -subj "/" -addext "subjectAltName = email:foo@example.com"
```

You can remove all the output below the command too, `-subj "/"` handles that so you aren't prompted.

You made need to update tests to check for this example SAN. Feel free to rerun the command too if you have a preferred email value.

File: pkg/cosign/verify.go
Comment (haydentherapper) [reply pkg/cosign/verify.go]: added comment

**Review by Hayden-IO — DISMISSED** [review]
LGTM, thanks so much for working on this!

**Review by Hayden-IO — DISMISSED** [review]

**Comment by haydentherapper** [comment]
@znewman01  once tests finish, do you want to merge?

**Comment by znewman01** [comment]
E2e tests are failing, once we get those fixed I can merge (will take a quick skim today too)

**Review by Hayden-IO — APPROVED** [review]
test failure is unrelated

**Review by znewman01 — APPROVED** [review]

**Comment by dmitris** [comment]
@znewman01 do you know if the docs on https://docs.sigstore.dev/cosign/keyless/ were updated for this change or not?

**Comment by haydentherapper** [comment]
Docs are being actively updated for cosign 2.0. Keyless will likely be removed as it’s not duplicated with the verify doc, which has been updated.

**Pull request merged**
